src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/ResourceScopeImpl.java

```diff
@@ -123,28 +123,43 @@ public static ResourceScopeImpl createShared(Cleaner cleaner) {
         return new SharedScope(cleaner);
     }
 
-    @Override
-    public final void release(Handle handle) {
+    private final void release0(HandleImpl handle) {
         try {
             Objects.requireNonNull(handle);
             if (!checkHandle(handle)) {
                 throw new IllegalArgumentException("Cannot release an handle acquired from another scope");
             }
-            ((HandleImpl) handle).close();
+            handle.close();
         } finally {
             Reference.reachabilityFence(this);
         }
     }
 
-    boolean checkHandle(Handle handle) {
+    @Override
+    public final void release(ResourceScope.Handle handle) {
+        release0((HandleImpl)handle);
+    }
+
+    @Override
+    public final void release(ScopedMemoryAccess.Scope.Handle handle) {
```

Comment:
Ugh - but I agree necessary

Reply:
Ugh! is right :-) I don't particularly like this, but its a temporary, private, necessary evil, which will eventually go away.